engine/src/main/java/org/pentaho/di/trans/steps/olapinput/OlapHelper.java

```diff
@@ -0,0 +1,191 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.trans.steps.olapinput;
+
+import java.sql.DriverManager;
+import java.sql.SQLException;
+
+import org.olap4j.CellSet;
+import org.olap4j.OlapConnection;
+import org.olap4j.OlapException;
+import org.olap4j.OlapStatement;
+import org.olap4j.OlapWrapper;
+import org.pentaho.di.core.exception.KettleDatabaseException;
+import org.pentaho.di.core.util.Utils;
+import org.pentaho.di.trans.steps.olapinput.olap4jhelper.AbstractBaseCell;
+import org.pentaho.di.trans.steps.olapinput.olap4jhelper.CellDataSet;
+import org.pentaho.di.trans.steps.olapinput.olap4jhelper.OlapUtil;
+
+/**
+ * <code>Olap4j Helper class</code> ...
+ *
+ * @author Paul Stoellberger
+ * @since Mar 12, 2010
+ */
+public class OlapHelper {
+
+  private String olap4jDriver;
+  private String olap4jUrl;
+  private String catalogName;
+  private String mdx;
+  private String username;
+  private String password;
+
+  private CellSet result;
+
+  private String[] headerValues = null;
+  private String[][] cellValues = null;
+  private OlapConnection olapConnection;
+
+  public OlapHelper( String olap4jDriver, String olap4jUrl, String username, String password, String catalogName,
+    String mdx ) {
+    this.olap4jDriver = olap4jDriver;
+    this.olap4jUrl = "jdbc:xmla:Server=" + olap4jUrl;
+    this.catalogName = catalogName;
+    this.mdx = mdx;
+    this.username = username;
+    this.password = password;
+  }
+
+  public void openQuery() throws Exception {
+
+    Class.forName( olap4jDriver );
+    OlapConnection connection = null;
+
+    if ( Utils.isEmpty( username ) && Utils.isEmpty( password ) ) {
+      connection = (OlapConnection) DriverManager.getConnection( olap4jUrl );
+    } else {
+      connection = (OlapConnection) DriverManager.getConnection( olap4jUrl, username, password );
+    }
+
+    OlapWrapper wrapper = connection;
+    olapConnection = wrapper.unwrap( OlapConnection.class );
+
+    try {
+      if ( !Utils.isEmpty( catalogName ) ) {
+        olapConnection.setCatalog( catalogName );
+      }
+    } catch ( SQLException e ) {
+      throw new OlapException( "Error setting catalog for MDX statement: '" + catalogName + "'" );
+    }
+
+    OlapStatement stmt = olapConnection.createStatement();
+
+    if ( !Utils.isEmpty( mdx ) ) {
+      CellSet tmp = stmt.executeOlapQuery( mdx );
+      result = tmp;
+    } else {
+      throw new Exception( "Error executing empty MDX query" );
+    }
+
+  }
+
+  public void close() throws KettleDatabaseException {
+    try {
+      if ( result != null ) {
+        result.close();
+      }
+      if ( olapConnection != null ) {
+        olapConnection.close();
+      }
+    } catch ( Exception e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)